main.go

```diff
@@ -3,117 +3,169 @@ package main
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"log"
+	"net"
+	"net/http"
 	"os"
 	"os/signal"
 	"path"
 	"syscall"
 
 	"github.com/hashicorp/vault-lambda-extension/config"
 	"github.com/hashicorp/vault-lambda-extension/extension"
+	"github.com/hashicorp/vault-lambda-extension/server"
 	"github.com/hashicorp/vault-lambda-extension/vault"
+	"github.com/hashicorp/vault/api"
 )
 
 const (
 	extensionName = "vault-lambda-extension"
 )
 
-var (
-	extensionClient = extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
-)
-
 func main() {
 	logger := log.New(os.Stderr, fmt.Sprintf("[%s] ", extensionName), log.Ldate|log.Ltime|log.LUTC)
 	ctx, cancel := context.WithCancel(context.Background())
+	extensionClient := extension.NewClient(os.Getenv("AWS_LAMBDA_RUNTIME_API"))
+	_, err := extensionClient.Register(ctx, extensionName)
+	if err != nil {
+		logger.Fatal(err)
+	}
+
+	threadFinishedCh := make(chan struct{}, 2)
```

Comment:
it may be worth documenting why this is a buffered channel of size 2, it's not immediately obvious to me and the double `<-threadFinishedCh` lower in this function isn't clear

Reply:
It might also be worth considering using a sync.WaitGroup instead of a channel for this.   It tends to make it easier to manage error cases, since the place where you signal completion (.Done) is closer to where you actually start a goroutine and call .Add.

Reply:
Good idea, done 👍